### PR TITLE
Pretrained flag and resnet50 pretrained weights

### PR DESCRIPTION
**Pull request by ananyahjha93**
## What does this PR do?

1. Change pretrained flag in ImageClassifier to load self-supervised pretrained weights for backbones
2. Add pretrained weights of simclr, swav, barlow-twins for resnet50 backbone from VISSL.
3. Removes dependency on bolts for pretrained SSL weights.

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,8 +10,13 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Added support for (input, target) style datasets (e.g. torchvision) to the from_datasets method ([#552](https://github.com/PyTorchLightning/lightning-flash/pull/552))
 
+- Added simclr, swav, barlow-twins pretrained weights for resnet50 backbone in ImageClassifier task ([#560](https://github.com/PyTorchLightning/lightning-flash/pull/560))
```
Comment (Borda) [inline CHANGELOG.md:13]: capitals for the names?

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,8 +10,13 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+
 ### Changed
 
+- Changed how pretrained flag works for loading weights for ImageClassifier task ([#560](https://github.com/PyTorchLightning/lightning-flash/pull/560))
```
Comment (Borda) [inline CHANGELOG.md:17]:
```suggestion
- Changed how pretrained flag works for loading weights for `ImageClassifier` task ([#560](https://github.com/PyTorchLightning/lightning-flash/pull/560))
```

File: flash/image/backbones.py [hunk 3]
```diff
@@ -71,27 +62,15 @@ def wrapper(*args, pretrained=False, **kwargs):
...
-
-
 if _TORCHVISION_AVAILABLE:
 
+    RESNET50_WEIGHTS_PATHS = {
+        "supervised": None,
+        "simclr": "https://dl.fbaipublicfiles.com/vissl/model_zoo/simclr_rn50_800ep_simclr_8node_resnet_16_07_20.7e8feed1/model_final_checkpoint_phase799.torch",
+        "swav": "https://dl.fbaipublicfiles.com/vissl/model_zoo/swav_in1k_rn50_800ep_swav_8node_resnet_27_07_20.a0a6b676/model_final_checkpoint_phase799.torch",
+        "barlow-twins": "https://dl.fbaipublicfiles.com/vissl/model_zoo/barlow_twins/barlow_twins_32gpus_4node_imagenet1k_1000ep_resnet50.torch",
+    }
```
Comment (Borda) [inline flash/image/backbones.py:72]:
```suggestion
    HTTPS_VISSL = "https://dl.fbaipublicfiles.com/vissl/model_zoo/"
    RESNET50_WEIGHTS_PATHS = {
        "supervised": None,
        "simclr": HTTPS_VISSL + "simclr_rn50_800ep_simclr_8node_resnet_16_07_20.7e8feed1/model_final_checkpoint_phase799.torch",
        "swav": HTTPS_VISSL + "swav_in1k_rn50_800ep_swav_8node_resnet_27_07_20.a0a6b676/model_final_checkpoint_phase799.torch",
        "barlow-twins": HTTPS_VISSL + "barlow_twins/barlow_twins_32gpus_4node_imagenet1k_1000ep_resnet50.torch",
    }
```

File: flash/image/backbones.py [hunk 4]
```diff
@@ -109,10 +88,40 @@ def _fn_mobilenet_vgg(model_name: str, pretrained: bool = True) -> Tuple[nn.Modu
             type=_type
         )
 
-    def _fn_resnet(model_name: str, pretrained: bool = True) -> Tuple[nn.Module, int]:
-        model: nn.Module = getattr(torchvision.models, model_name, None)(pretrained)
+    def _fn_resnet(model_name: str,
+                   pretrained: Union[bool, str] = True,
+                   weights_paths: dict = {"supervised": None}) -> Tuple[nn.Module, int]:
```
Comment (Borda) [inline flash/image/backbones.py:93]:
```suggestion
                   weights_paths: dict = {"supervised": None},) -> Tuple[nn.Module, int]:
```

File: flash/image/backbones.py [hunk 5]
```diff
@@ -109,10 +88,40 @@ def _fn_mobilenet_vgg(model_name: str, pretrained: bool = True) -> Tuple[nn.Modu
...
+        # load according to pretrained if a bool is specified, else set to False
+        if isinstance(pretrained, bool):
+            pretrained_flag = pretrained
+        else:
+            pretrained_flag = False
```
Comment (Borda) [inline flash/image/backbones.py:98]:
```suggestion
        pretrained_flag = pretrained isinstance(pretrained, bool) else False
```
increase coverage

File: flash/image/backbones.py [hunk 6]
```diff
@@ -109,10 +88,40 @@ def _fn_mobilenet_vgg(model_name: str, pretrained: bool = True) -> Tuple[nn.Modu
...
+        if isinstance(pretrained, str):
+            if pretrained in weights_paths:
+                device = next(model.parameters()).get_device()
+                model_weights = load_state_dict_from_url(
+                    weights_paths[pretrained],
+                    map_location=torch.device('cpu') if device is -1 else torch.device(device)
+                )
+
+                # add logic here for loading resnet weights from other libraries
+                if "classy_state_dict" in model_weights.keys():
```
Comment (Borda) [inline flash/image/backbones.py:114]:
```suggestion
                if "classy_state_dict" in model_weights:
```

File: flash/image/backbones.py [hunk 7]
```diff
@@ -109,10 +88,40 @@ def _fn_mobilenet_vgg(model_name: str, pretrained: bool = True) -> Tuple[nn.Modu
...
+                    model_weights = model_weights["classy_state_dict"]["base_model"]["model"]["trunk"]
```
Comment (Borda) [inline flash/image/backbones.py:115]:
```suggestion
                    model_weights_ = model_weights["classy_state_dict"]["base_model"]["model"]["trunk"]
```
rather another var?

File: flash/image/backbones.py [hunk 8]
```diff
@@ -125,13 +134,24 @@ def _fn_resnet_fpn(
...
+            )
+        else:
+            IMAGE_CLASSIFIER_BACKBONES(
+                fn=catch_url_error(partial(_fn_resnet, model_name=model_name)),
+                name=model_name,
+                namespace="vision",
+                package="torchvision",
+                type="resnet",
+                weights_paths={"supervised": None}
+            )
```
Comment (Borda) [inline flash/image/backbones.py:154]:
```suggestion
        clf_kwargs = dict(
            fn=catch_url_error(partial(_fn_resnet, model_name=model_name)),
            name=model_name,
            namespace="vision",
            package="torchvision",
            type="resnet",
            weights_paths={"supervised": None}
        )
        if model_name == 'resnet50':
            clf_kwargs.update(dict(
                fn=catch_url_error(partial(_fn_resnet, model_name=model_name, weights_paths=RESNET50_WEIGHTS_PATHS)),
                package="multiple",
                weights_paths=RESNET50_WEIGHTS_PATHS
            )
        IMAGE_CLASSIFIER_BACKBONES(**clf_kwargs)
```

**Review by Borda — APPROVED** [review]

File: flash/image/backbones.py [hunk 9]
```diff
@@ -109,10 +89,37 @@ def _fn_mobilenet_vgg(model_name: str, pretrained: bool = True) -> Tuple[nn.Modu
             type=_type
         )
 
-    def _fn_resnet(model_name: str, pretrained: bool = True) -> Tuple[nn.Module, int]:
-        model: nn.Module = getattr(torchvision.models, model_name, None)(pretrained)
+    def _fn_resnet(model_name: str,
+                   pretrained: Union[bool, str] = True,
+                   weights_paths: dict = {"supervised": None},) -> Tuple[nn.Module, int]:
+        # load according to pretrained if a bool is specified, else set to False
+        pretrained_flag = pretrained isinstance(pretrained, bool) else False
```
Comment (ethanwharris) [inline flash/image/backbones.py:96]:
```suggestion
        pretrained_flag = (pretrained and isinstance(pretrained, bool)) or (pretrained == "supervised")
```

File: flash/image/backbones.py [hunk 10]
```diff
@@ -109,10 +89,37 @@ def _fn_mobilenet_vgg(model_name: str, pretrained: bool = True) -> Tuple[nn.Modu
...
+
+        model: nn.Module = getattr(torchvision.models, model_name, None)(pretrained_flag)
         backbone = nn.Sequential(*list(model.children())[:-2])
         num_features = model.fc.in_features
+
+        model_weights = None
+        if isinstance(pretrained, str):
```
Comment (ethanwharris) [inline flash/image/backbones.py:103]:
```suggestion
        if not pretrained_flag and isinstance(pretrained, str):
```

File: flash/image/backbones.py [hunk 11]
```diff
@@ -109,10 +89,37 @@ def _fn_mobilenet_vgg(model_name: str, pretrained: bool = True) -> Tuple[nn.Modu
...
+                if "classy_state_dict" in model_weights:
+                    model_weights = model_weights["classy_state_dict"]["base_model"]["model"]["trunk"]
+                    model_weights = {
+                        key.replace("_feature_blocks.", "") if "_feature_blocks." in key else key: val
+                        for (key, val) in model_weights.items()
+                    }
+                else:
+                    raise KeyError('Model state dict loaded from unrecognized url/path.')
+            else:
+                raise KeyError('Unrecognized pretrained model.')
```
Comment (ethanwharris) [inline flash/image/backbones.py:121]:
```suggestion
```

File: flash/image/backbones.py [hunk 12]
```diff
@@ -109,10 +89,37 @@ def _fn_mobilenet_vgg(model_name: str, pretrained: bool = True) -> Tuple[nn.Modu
...
+
+                # add logic here for loading resnet weights from other libraries
+                if "classy_state_dict" in model_weights:
+                    model_weights = model_weights["classy_state_dict"]["base_model"]["model"]["trunk"]
+                    model_weights = {
+                        key.replace("_feature_blocks.", "") if "_feature_blocks." in key else key: val
+                        for (key, val) in model_weights.items()
+                    }
+                else:
+                    raise KeyError('Model state dict loaded from unrecognized url/path.')
```
Comment (ethanwharris) [inline flash/image/backbones.py:119]: Let's give a list of available weights paths here from `weights_paths.keys()`

**Review by ethanwharris — COMMENTED** [review]
Looking good :smiley: I think there's some slightly faulty logic in the load function, maybe some tests for different `pretrained` configurations would be good

**Comment by ananyahjha93** [comment]
@ethanwharris is there a way to cache the pretrained weights? Adding tests for pretrained configurations will add significant time to tests since large models need to be downloaded before testing. Hence I wasn't sure to include this.

**Comment by ethanwharris** [comment]
@ananyahjha93 I think they will get cached automatically but not sure. We should still unit test the function though, you could use `patch` to prevent things from being downloaded

**Review by ethanwharris — APPROVED** [review]
Awesome, LGTM :smiley:

**Pull request merged**